### DISCUSSION TITLE
Question around getClassesWithAnnotation() and classes outside the scanSpec.

### DISCUSSION THREAD

**Discussion by ginkoblongata** [classgraph/classgraph#697] [Q&A]
Hello, I am pretty excited about your project.

I have a clarification of usage question because I am seemingly getting strange results and perhaps I have not delved into the api enough yet.

Firstly, I will show my setup.  Secondly, I will describe how I'm having unexpected results.

```
ScanResult scanResult = new ClassGraph().enableAllInfo().acceptPackages("XYZ").scan();
List all = scanResult.getAllClasses();
System.out.println("here are all: " + all);

List subset = scanResult.getClassesWithAnnotation(Configuration.class.getName());
System.out.println("here should be subset: " + subset);
```

The issue I am seeing is there are 3 classes showing up in the 'subset' list which are not present in the 'all' list.
Also, the classes that show up in the subset list are not within the single supplied base package "XYZ", they are in "ABC" which is not scanned for.

Is there something about getClassesWithAnnotation() which I would need to better understand (like expected side effects), or is this the correct usage?

Thanks!

===============

My current thinking is that getClassesWithAnnotation() does not mean "filter getAllClasses by annotation" but rather means all classes (which includes fields and parameters) encountered by the scan of classes in the base-package.

In this case, then I would call getAllClasses() and then do the filtering.
Is this the correct approach?

Thanks!

**Comment by lukehutch** [comment]
Hi, are the extra annotations meta-annotations, ie. do they annotate annotation classes in the scanned packages? See #559.

**Comment by ginkoblongata** [comment]
Hi,

That issue was interesting, but in this case, these are pretty bland classes.
I did take the other approach to just call getAllClasses() and do the filtering on my end and it is working, so I think it is something going on with the wording of the comments and the name of the method and my interpretation not matching up with the intention.

I have to go back and see about making a reproduction since I've moved on with the code.

But now I am seeing a situation where if I do not add a package to the scan, then I get null for some ClassInfo objects, it was either parameter or return types, I'll have to check that too.

But I am seeing a method called enableExternalClasses() and it seems that it should be included in the ClassGraph.enableAllInfo() method.
However, when I add that to the scan, it does not alleviate the unwanted need to add the additional packages to the scan.
I see there is also enableInterClassDependencies(), I sill have to play around with those some to see if I can get a correct usage.

But as it stands, it seems they are all related, I am able to get the correct results but with a wrinkle.
When I do not add the extra packages to the scan, then I get the unwanted results in the getClassesWithAnnotation() method from what I would consider 'external' classes, which is before I was aware of those various methods, but I have stopped using that.
When I proceed further, I get null pointers for ClassInfo when I try to determine the define type for either parameter or return type. (I forget which)
So then I add the packages, and then things work but I start noticing these 'external' methods which don't appear to interact with this, but again I have to try those out more I guess.
